### DISCUSSION TITLE
template x-for not display order correctly, after drag elemnts with Sortable JS

### DISCUSSION THREAD

**Discussion by javiabellan** [alpinejs/alpine#4368] [1. Help] [answered]
I want to reflect to state after an element is drag and droped with [Sortable JS](https://github.com/SortableJS/Sortable) lib into the Alpine JS state:

I can achieve that by using the Sortable JS callback `onEnd` to update my Alpine JS state. The problem is afterwards, Alpine JS do a re-rendeing that breaks the order.

Here is my complete code:

`index.html`:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Sortable List Example</title>

    <script type="text/javascript" src="logic.js" defer></script>

    <!-- TAILWIND & ALPINE JS -->
    <script src="https://cdn.tailwindcss.com"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.14.1/dist/cdn.min.js"></script>

    <style type="text/css">
        .sortable-ghost { opacity: 0.2; }
    </style>
  </head>

  <body x-data="sortableList()">
    <div x-ref="list" class="flex rounded-lg my-16 mx-32 border-2">
      <template x-for="(item, index) in items" :key="index"> <!--:key="item.id"-->
        <div x-text="item.name"
        class="flex rounded bg-slate-200 m-2 p-2 border-2 border-slate-500">
        </div>
      </template>
    </div>

    <!-- Load the SortableJS library -->
    <script src="https://cdn.jsdelivr.net/npm/sortablejs@latest/Sortable.min.js"></script>

  </body>
</html>
```

`logic.js`

```js
function sortableList() {
  return {
    items: [
      { name: 'Item 1' },
      { name: 'Item 2' },
      { name: 'Item 3' },
      { name: 'Item 4' },
      { name: 'Item 5' },
    ],
    init() {
      console.log("init");
      let el = this.$refs.list;
      new Sortable(el, {
        animation: 200,

        onEnd: (evt) => {
          const {newIndex, oldIndex} = evt;
          console.log(oldIndex, "-->", newIndex);
          if (oldIndex !== newIndex) {
            const movedItem = this.items.splice(oldIndex, 1)[0];
            this.items.splice(newIndex, 0, movedItem);
            
            // Here we check that the items sorting update logic is performed properly
            console.log(this.items);
          }
        }
      });
    }
  }
}
```

If the `onEnd` callback is removed, the drag & drop interface works as expected, but there is no updating into the alpine JS `items` state.

I would like to find what is the best coding pattern, where the interface changes the Alpine state, instead of the common pattern where Alpine state changes the interface.

Thanks in advance.

**Comment by SimoTod** [comment]
Don't use an index as key for your x-for item. The key should be something that identify your element and doesn't change between rendering. You can use item.name, assuming they are unique.

**Reply by javiabellan** [reply]
that's does not solve the problem:

I tried:

```js
    items: [
      { id: 1, name: 'Item 1' },
      { id: 2, name: 'Item 2' },
      { id: 3, name: 'Item 3' },
      { id: 4, name: 'Item 4' },
      { id: 5, name: 'Item 5' },
    ],
```

with

```html
<template x-for="(item, index) in items" :key="item.id">
```

and the problem persists.

**Comment by javiabellan** [comment]
If you execute the code, you can interact with it.

For example if you drag the "item 5" before the "item 4". You will see that that state is changed properly, but the rendering fails:
<img width="1328" alt="Captura de pantalla 2024-09-14 a las 15 21 01" src="https://github.com/user-attachments/assets/f7dc34da-ad6c-44a1-b1f0-6374730375a9">

**Reply by SimoTod** [reply]
Try calling Alpine.deferMutations before the first splice call and Alpine.flushAndStopDeferringMutations() after the second one.
If it doesn't work, I'll send you a working example in a few hours when I'm at my laptop.

**Reply by javiabellan** [reply]
Thanks a lot, I really appreciate it. I'm going to try that!

**Reply by javiabellan** [reply]
Sadly the deferMutations wrapper did not solve the problem.

```js
onEnd: (evt) => {
  const {newIndex, oldIndex} = evt;
  if (oldIndex !== newIndex) {
    Alpine.deferMutations();
    const movedItem = this.items.splice(oldIndex, 1)[0];
    this.items.splice(newIndex, 0, movedItem);
    Alpine.flushAndStopDeferringMutations();
  }
}
```

If you could find a working example i would really appreciate it. Thanks in advance @SimoTod ! :)

**Reply by javiabellan** [reply, marked as answer]
I have found the solution at [this previous discussion](https://github.com/alpinejs/alpine/discussions/1635)

I just need it to write `const items = Alpine.raw(this.items);` before the first `splice`. However i am not sure why it works, im not familiar with `Alpine.raw`.

```js
onEnd: (evt) => {
  const {newIndex, oldIndex} = evt;
  if (oldIndex !== newIndex) {
    const items = Alpine.raw(this.items); // <----- THIS SOLVES THE PROBLEM
    const movedItem = items.splice(oldIndex, 1)[0]; // Then i use items instead of this.items
    items.splice(newIndex, 0, movedItem);
  }
}
```

**Reply by SimoTod** [reply]
It doesn't completely solves the problem (unless your x-for is the only directive accessing the list) but it may be okay for your use case as long as you understand what it does.
When you use Alpine.raw, you are getting the real value without the reactivity layer Alpine uses to react to your changes. 
As a consequence, if you have any other directives using items, they won't know about the change and will not re-render. Your x-for won't re-render either which is why it fixes your problem (when re rendering, Alpine would update the list, not the redesigning the original one, but those elements have been moved by SortableJS and are not in the position Alpine thinks they should be). See https://codepen.io/SimoTod/pen/qBzGZmL

It may be okay for your use case, but I thought it was worth pointing it out in case someone else finds this post in the future and have a slightly different requirement.

The problem for you is that Alpine does update the elements but those have been moved to a different place by SortableJs resulting in a double swap because Alpine is trying to reuse existing elements and it doesn't know that they are not in the same position anymore. Perhaps, in your case, the opposite approach could be the solution (unless you have hundreds of items since it will be slow on bug numbers), you can pick a key that changes every time so Alpine will never reuse existing element but only recreate the item at every re-rendering. See https://codepen.io/SimoTod/pen/LYKoNro

**Reply by SimoTod** [reply]
Also, it's worth noting that https://github.com/alpinejs/alpine/pull/4361 is in review and when it gets merged, it may fix your use case without workarounds.

**Reply by javiabellan** [reply]
Thanks @SimoTod for explaining in detail what is happening under the hood. I appreciate it.

Yes, `Alpine.raw` solves my problem because i only render the list once. Thanks for pointing that!

**Reply by ekwoka** [reply]
Yeah, `raw` working here is, to my eyes, more of a happy accident than a stable solution.

Thinking it through I can't remotely imagine WHY that would change anything at all...

Though, also in that, I can't imagine any reason why it being there would cause anything to break in the future either :thinking:
